src/box/lua/schema.lua

```diff
@@ -3453,35 +3470,81 @@ local function grant(uid, name, privilege, object_type,
     else
         options.grantor = user_or_role_resolve(options.grantor)
     end
+    if options._origin == nil then
+        options._origin = DEFAULT_ORIGIN
+    elseif type(options._origin) ~= 'string' then
+        box.error(box.error.ILLEGAL_PARAMS, "options parameter '_origin' " ..
+                  "should be of type 'string'")
+    end
     local _priv = box.space[box.schema.PRIV_ID]
     local _vpriv = box.space[box.schema.VPRIV_ID]
     -- add the granted privilege to the current set
     local tuple = _vpriv:get{uid, object_type, oid}
-    local old_privilege
-    if tuple ~= nil then
-        old_privilege = tuple.privilege
+
+    local origins
+    if tuple == nil then
+        origins = {[DEFAULT_ORIGIN] = 0}
+    elseif tuple[PRIV_OPTS_FIELD_ID] == nil or
+           tuple[PRIV_OPTS_FIELD_ID].origins == nil then
+        origins = {[DEFAULT_ORIGIN] = tuple.privilege}
     else
-        old_privilege = 0
+        origins = tuple[PRIV_OPTS_FIELD_ID].origins
     end
-    privilege_hex = bit.bor(privilege_hex, old_privilege)
-    privilege_hex = tonumber(ffi.cast('uint32_t', privilege_hex))
+
     -- do not execute a replace if it does not change anything
     -- XXX bug if we decide to add a grant option: new grantor
     -- replaces the old one, old grantor is lost
-    if privilege_hex ~= old_privilege then
-        _priv:replace{options.grantor, uid, object_type, oid, privilege_hex}
-    elseif not options.if_not_exists then
-            if object_type == 'role' and object_name ~= '' and
-               privilege == 'execute' then
-                box.error(box.error.ROLE_GRANTED, name, object_name)
-            else
-                if not is_singleton_object_type(object_type) then
-                    object_name = string.format(" '%s'", object_name)
-                end
-                box.error(box.error.PRIV_GRANTED, name, privilege,
-                          object_type, object_name)
-            end
+    local old_privilege = origins[options._origin] or 0
+    if bit.band(bit.bnot(old_privilege), privilege_hex) ~= 0 then
+        -- Update given privileges by origin.
+        origins[options._origin] = bit.bor(old_privilege, privilege_hex)
+    else
+        -- No new privileges can be given.
+        if options.if_not_exists then
+            return
+        end
+        grant_error(name, object_name, object_type, privilege)
+    end
+
+    -- Total privileges.
+    local new_privilege = 0
+    for origin_name, origin_privilege in pairs(origins) do
+        new_privilege = bit.bor(new_privilege, origin_privilege)
+        if origin_privilege == 0 then
+            origins[origin_name] = nil
+        end
     end
+    local res = {options.grantor, uid, object_type, oid, new_privilege}
+    -- If only default origin present, we do not need opts.
+    assert(next(origins) ~= nil)
+    if origins[DEFAULT_ORIGIN] == nil or
+       next(origins, next(origins)) ~= nil then
```

Comment:
We have such a code as in `grant` as well as in `revoke`. Can we extract it into its own functions?

For example, `privilege_from_origins(origins)` and `normalize_origins(origins)`. The latter should strip origins with zero privileges and should return `nil` if origins shouldn't be stored (equals to `{[''] = new_privilege}` or just has only `''` key).

Reply:
Done. Introduced two new functions: `privilege_from_origins()` and `opts_from_origins()`.